src/test/java/org/folio/marc4ld/service/marc2ld/postprocessor/InstanceResourcePostProcessorTest.java

```diff
@@ -0,0 +1,112 @@
+
+  @Test
+  void shouldCleanEmptyEdges() {
+    // given
+    var emptyResource = createResource(1L, null);
+    var nonEmptyResource1 = createResource(2L, OBJECT_MAPPER.createObjectNode().put("k", "v"));
+    var nonEmptyResource2 = createResource(3L, null, nonEmptyResource1);
+    var testResource = createResource(4L, null, nonEmptyResource1, nonEmptyResource2, emptyResource);
+
+    // when
+    instanceResourcePostProcessor.apply(testResource);
+
+    // then
+    assertThat(testResource.getOutgoingEdges()).hasSize(2);
+    assertThat(testResource.getOutgoingEdges().stream().map(ResourceEdge::getTarget))
+      .containsExactlyInAnyOrder(nonEmptyResource1, nonEmptyResource2);
+  }
+
+  @Test
+  void shouldAssignIdRecursively() {
+    // given
+    var resource1 = createResource(1L, OBJECT_MAPPER.createObjectNode().put("k", "v"));
+    var resource2 = createResource(2L, null, resource1);
+    var resource3 = createResource(3L, null, resource1);
+    var testResource = createResource(4L, null, resource2, resource3);
+    when(fingerprintHashService.hash(resource1)).thenReturn(100L);
+    when(fingerprintHashService.hash(resource2)).thenReturn(200L);
+    when(fingerprintHashService.hash(resource3)).thenReturn(300L);
+    when(fingerprintHashService.hash(testResource)).thenReturn(400L);
+
+    // when
+    instanceResourcePostProcessor.apply(testResource);
+
+    // then
+    assertThat(resource1.getId()).isEqualTo(100L);
+    assertThat(resource2.getId()).isEqualTo(200L);
+    assertThat(resource3.getId()).isEqualTo(300L);
+    assertThat(testResource.getId()).isEqualTo(400L);
+  }
+
+  @Test
+  void shouldSetInstanceLabel() {
+    // given
+    var expectedLabel = "instance title";
+    var instance = createResource(2L, null);
+    var title = createResource(3L, null).setLabel(expectedLabel);
+    createConnection(instance, title, TITLE);
+
+    // when
+    instanceResourcePostProcessor.apply(instance);
+
+    // then
+    assertThat(instance.getLabel()).isEqualTo(expectedLabel);
+  }
+
+  @Test
+  void shouldSetWorkLabel() {
+    // given
+    var expectedLabel = "work title";
+    var title = createResource(3L, null).setLabel(expectedLabel);
+    var instance = createResource(2L, null);
+    var work = createResource(3L, null);
+    createConnection(instance, title, TITLE);
+    createConnection(instance, work, INSTANTIATES);
+
+    // when
+    instanceResourcePostProcessor.apply(instance);
+
+    // then
+    assertThat(work.getLabel()).isEqualTo(expectedLabel);
+  }
+
+  private Resource createResource(Long id, JsonNode doc, Resource... edges) {
+    var resource = new Resource().setId(id).setDoc(doc);
+    Stream.of(edges)
+      .forEach(e -> createConnection(resource, e, EDITOR));
+    return resource;
+  }
+
+  private void createConnection(Resource from, Resource to, PredicateDictionary predicate) {
+    from.getOutgoingEdges().add(new ResourceEdge(from, to, predicate));
```

Comment:
Minor - Looks like we need to add an `addOutgoingEdge(..)` method in the Resource class.

Reply:
sure, why not!